Diff snapshot requires two test runs for new tests and two to 'stabilise' after changes to base snapshot

**Describe the bug**

Currently, using `diff` seems to take two test runs with `--snapshot-update` to have the diff computed correctly in two cases:

- if a test is newly added with both a base snapshot and a diff, the diff snapshot will fail with a `SnapshotDoesNotExist` error on the first run, but work on the second
- if the "base"/original snapshot has a change, the first test run will have those changes appear in the diff, and a second run is required to clear them out


**To reproduce**

For a new test:

1. Create a new test like the following:
    ```python
    base = {"A": 1}
    def test_foo(snapshot):
        assert {**base} == snapshot(name="a")
        assert {**base, "B": 2} == snapshot(name="b", diff="a")
    ```
2. run `pytest --snapshot-update`
3. BUG: observe it fails with an error like:
    ```
    >       assert {**base, "B": 2} == snapshot(name="b", diff="a")
    E       assert [+ received] == [- snapshot]
    E         syrupy.exceptions.SnapshotDoesNotExist
    E         Traceback (most recent call last):
    E           File "/private/var/folders/sv/vd266m4d4lvctgs2wpnhjs9w0000gn/T/tmp.Kn7Y7pGLWR/.venv/lib/python3.10/site-packages/syrupy/assertion.py", line 322, in _assert
    E             raise SnapshotDoesNotExist()
    
    test_foo.py:4: AssertionError
    ```
4. run `pytest --snapshot-update` again (no other changes) and observe it works

For an existing test:

5. Take the test above (after running `pytest --snapshot-update` twice) and edit `base` to have an extra key, `base = {"A": 1, "C": 3}`
6. Run `pytest --snapshot-update`
7. BUG: observe the `__snapshot__/test_foo.ambr` file contains the following, and, in particular, the diff says that `'C': 3` was added... but it appears in the `a` snapshot too, i.e. appears in both snapshots and thus shouldn't be in the diff!
    ```
    # serializer version: 1
    # name: test_foo[a]
      dict({
        'A': 1,
        'C': 3,
      })
    # ---
    # name: test_foo[b]
          ...
      +   'B': 2,
      +   'C': 3,
        ...
    # ---
    ```
8. Run `pytest` (no `--snapshot-update`) and observe failures:
    ```
    >       assert {**base, "B": 2} == snapshot(name="b", diff="a")
    E       AssertionError: assert [+ received] == [- snapshot]
    E               ...
    E           +   'B': 2,
    E         - +   'C': 3,
    E             ...
    
    test_foo.py:4: AssertionError
    ```

An all-in-one shell script reproducer: save it and run with `bash ./script.sh`

<details>

```shell
cd $(mktemp -d)

python -m venv .venv
. .venv/bin/activate

pip install syrupy==4.8.0 colored==1.4.4 exceptiongroup==1.2.2 iniconfig==2.0.0 packaging==24.2 pluggy==1.5.0 pytest==7.4.4 tomli==2.2.1

cat > test_foo.py <<EOF
base = {"A": 1}
def test_foo(snapshot):
    assert {**base} == snapshot(name="a")
    assert {**base, "B": 2} == snapshot(name="b", diff="a")
EOF

echo '**** BUG: first run fails with `raise SnapshotDoesNotExist`'
pytest test_foo.py --snapshot-update
echo '**** OKAY: second run passes'
pytest test_foo.py --snapshot-update

# add a new field C to base
cat > test_foo.py <<EOF
base = {"A": 1, "C": 3}
def test_foo(snapshot):
    assert {**base} == snapshot(name="a")
    assert {**base, "B": 2} == snapshot(name="b", diff="a")
EOF

echo '**** BUG: running after a change has the diff snapshot think C is added (should only be B)'
pytest test_foo.py --snapshot-update
cat __snapshots__/test_foo.ambr
echo '**** ... which means a second non-snapshot test run fails'
pytest test_foo.py
```

</details>

**Expected behavior**

Running `pytest --snapshot-update` should make the complete set of updates required, i.e. assuming the snapshots are reproducible, the both:

- running `pytest --snapshot-update` a second time should make no changes
- running `pytest` alone should pass


**Screenshots**

<!-- If applicable, add screenshots to help explain your problem. -->

**Environment (please complete the following information):**

- OS: macOS 15
- Syrupy Version: 4.8
- Python Version: 3.10

**Additional context**

Maybe very slightly related to https://github.com/syrupy-project/syrupy/issues/608

(Thanks for syrupy! :pancakes: )

--- a/src/syrupy/assertion.py
+++ b/src/syrupy/assertion.py
@@ -377,11 +377,7 @@ class SnapshotAssertion:
     ) -> Tuple[Optional["SerializableData"], bool]:
         try:
             return (
-                self.extension.read_snapshot(
-                    test_location=self.test_location,
-                    index=index,
-                    session_id=str(id(self.session)),
-                ),
+                self.session.recall_snapshot(self.extension, self.test_location, index),
                 False,
             )
         except SnapshotDoesNotExist:

--- a/src/syrupy/location.py
+++ b/src/syrupy/location.py
@@ -13,7 +13,7 @@ import pytest
 from syrupy.constants import PYTEST_NODE_SEP
 
 
-@dataclass
+@dataclass(frozen=True)
 class PyTestLocation:
     item: "pytest.Item"
     nodename: Optional[str] = field(init=False)
@@ -23,27 +23,41 @@ class PyTestLocation:
     filepath: str = field(init=False)
 
     def __post_init__(self) -> None:
+        # NB. we're in a frozen dataclass, but need to transform the values that the caller
+        # supplied... we do so by (ab)using object.__setattr__ to forcibly set the attributes. (See
+        # rejected PEP-0712 for an example of a better way to handle this.)
+        #
+        # This is safe because this all happens during initialization: `self` hasn't been hashed
+        # (or, e.g., stored in a dict), so the mutation won't be noticed.
         if self.is_doctest:
             return self.__attrs_post_init_doc__()
         self.__attrs_post_init_def__()
 
     def __attrs_post_init_def__(self) -> None:
         node_path: Path = getattr(self.item, "path")  # noqa: B009
-        self.filepath = str(node_path.absolute())
+        # See __post_init__ for discussion of object.__setattr__
+        object.__setattr__(self, "filepath", str(node_path.absolute()))
         obj = getattr(self.item, "obj")  # noqa: B009
-        self.modulename = obj.__module__
-        self.methodname = obj.__name__
-        self.nodename = getattr(self.item, "name", None)
-        self.testname = self.nodename or self.methodname
+        object.__setattr__(self, "modulename", obj.__module__)
+        object.__setattr__(self, "methodname", obj.__name__)
+        object.__setattr__(self, "nodename", getattr(self.item, "name", None))
+        object.__setattr__(self, "testname", self.nodename or self.methodname)
 
     def __attrs_post_init_doc__(self) -> None:
         doctest = getattr(self.item, "dtest")  # noqa: B009
-        self.filepath = doctest.filename
+        # See __post_init__ for discussion of object.__setattr__
+        object.__setattr__(self, "filepath", doctest.filename)
         test_relfile, test_node = self.nodeid.split(PYTEST_NODE_SEP)
         test_relpath = Path(test_relfile)
-        self.modulename = ".".join([*test_relpath.parent.parts, test_relpath.stem])
-        self.nodename = test_node.replace(f"{self.modulename}.", "")
-        self.testname = self.nodename or self.methodname
+        object.__setattr__(
+            self,
+            "modulename",
+            ".".join([*test_relpath.parent.parts, test_relpath.stem]),
+        )
+        object.__setattr__(
+            self, "nodename", test_node.replace(f"{self.modulename}.", "")
+        )
+        object.__setattr__(self, "testname", self.nodename or self.methodname)
 
     @property
     def classname(self) -> Optional[str]:

--- a/src/syrupy/location.py
+++ b/src/syrupy/location.py
@@ -54,6 +54,7 @@ class PyTestLocation:
             "modulename",
             ".".join([*test_relpath.parent.parts, test_relpath.stem]),
         )
+        object.__setattr__(self, "methodname", None)
         object.__setattr__(
             self, "nodename", test_node.replace(f"{self.modulename}.", "")
         )

--- a/src/syrupy/session.py
+++ b/src/syrupy/session.py
@@ -64,7 +64,7 @@ class SnapshotSession:
 
     _queued_snapshot_writes: Dict[
         Tuple[Type["AbstractSyrupyExtension"], str],
-        List[Tuple["SerializedData", "PyTestLocation", "SnapshotIndex"]],
+        Dict[Tuple["PyTestLocation", "SnapshotIndex"], "SerializedData"],
     ] = field(default_factory=dict)
 
     def _snapshot_write_queue_key(
@@ -86,8 +86,8 @@ class SnapshotSession:
         index: "SnapshotIndex",
     ) -> None:
         key = self._snapshot_write_queue_key(extension, test_location, index)
-        queue = self._queued_snapshot_writes.get(key, [])
-        queue.append((data, test_location, index))
+        queue = self._queued_snapshot_writes.get(key, {})
+        queue[(test_location, index)] = data
         self._queued_snapshot_writes[key] = queue
 
     def flush_snapshot_write_queue(self) -> None:
@@ -97,7 +97,11 @@ class SnapshotSession:
         ), queued_write in self._queued_snapshot_writes.items():
             if queued_write:
                 extension_class.write_snapshot(
-                    snapshot_location=snapshot_location, snapshots=queued_write
+                    snapshot_location=snapshot_location,
+                    snapshots=[
+                        (data, loc, index)
+                        for (loc, index), data in queued_write.items()
+                    ],
                 )
         self._queued_snapshot_writes = {}
 
@@ -110,12 +114,10 @@ class SnapshotSession:
         """Find the current value of the snapshot, for this session, either a pending write or the actual snapshot."""
 
         key = self._snapshot_write_queue_key(extension, test_location, index)
-        queue = self._queued_snapshot_writes.get(key)
-        if queue:
-            # find the last (i.e. most recent) write to this index/location in the queue:
-            for queue_data, queue_test_location, queue_index in reversed(queue):
-                if queue_index == index and queue_test_location == test_location:
-                    return queue_data
+        queue = self._queued_snapshot_writes.get(key, {})
+        data = queue.get((test_location, index))
+        if data is not None:
+            return data
 
         # no queue, or no matching write, so just read the snapshot directly:
         return extension.read_snapshot(

--- a/src/syrupy/session.py
+++ b/src/syrupy/session.py
@@ -66,10 +66,10 @@ class SnapshotSession:
         default_factory=lambda: defaultdict(set)
     )
 
-    _queued_snapshot_writes: Dict[
+    _queued_snapshot_writes: DefaultDict[
         _QueuedWriteExtensionKey,
         Dict[_QueuedWriteTestLocationKey, "SerializedData"],
-    ] = field(default_factory=dict)
+    ] = field(default_factory=lambda: defaultdict(dict))
 
     def _snapshot_write_queue_keys(
         self,
@@ -92,9 +92,7 @@ class SnapshotSession:
         ext_key, loc_key = self._snapshot_write_queue_keys(
             extension, test_location, index
         )
-        queue = self._queued_snapshot_writes.get(ext_key, {})
-        queue[loc_key] = data
-        self._queued_snapshot_writes[ext_key] = queue
+        self._queued_snapshot_writes[ext_key][loc_key] = data
 
     def flush_snapshot_write_queue(self) -> None:
         for (
@@ -109,7 +107,7 @@ class SnapshotSession:
                         for (loc, index), data in queued_write.items()
                     ],
                 )
-        self._queued_snapshot_writes = {}
+        self._queued_snapshot_writes.clear()
 
     def recall_snapshot(
         self,
@@ -122,8 +120,7 @@ class SnapshotSession:
         ext_key, loc_key = self._snapshot_write_queue_keys(
             extension, test_location, index
         )
-        queue = self._queued_snapshot_writes.get(ext_key, {})
-        data = queue.get(loc_key)
+        data = self._queued_snapshot_writes[ext_key].get(loc_key)
         if data is not None:
             return data
 

--- a/src/syrupy/session.py
+++ b/src/syrupy/session.py
@@ -66,6 +66,13 @@ class SnapshotSession:
         default_factory=lambda: defaultdict(set)
     )
 
+    # For performance, we buffer snapshot writes in memory before flushing them to disk. In
+    # particular, we want to be able to write to a file on disk only once, rather than having to
+    # repeatedly rewrite it.
+    #
+    # That batching leads to using two layers of dicts here: the outer layer represents the
+    # extension/file-location pair that will be written, and the inner layer represents the
+    # snapshots within that, "indexed" to allow efficient recall.
     _queued_snapshot_writes: DefaultDict[
         _QueuedWriteExtensionKey,
         Dict[_QueuedWriteTestLocationKey, "SerializedData"],
@@ -124,7 +131,7 @@ class SnapshotSession:
         if data is not None:
             return data
 
-        # no queue, or no matching write, so just read the snapshot directly:
+        # No matching write queued, so just read the snapshot directly:
         return extension.read_snapshot(
             test_location=test_location, index=index, session_id=str(id(self))
         )

--- a/src/syrupy/session.py
+++ b/src/syrupy/session.py
@@ -67,6 +67,17 @@ class SnapshotSession:
         List[Tuple["SerializedData", "PyTestLocation", "SnapshotIndex"]],
     ] = field(default_factory=dict)
 
+    def _snapshot_write_queue_key(
+        self,
+        extension: "AbstractSyrupyExtension",
+        test_location: "PyTestLocation",
+        index: "SnapshotIndex",
+    ) -> Tuple[Type["AbstractSyrupyExtension"], str]:
+        snapshot_location = extension.get_location(
+            test_location=test_location, index=index
+        )
+        return (extension.__class__, snapshot_location)
+
     def queue_snapshot_write(
         self,
         extension: "AbstractSyrupyExtension",
@@ -74,10 +85,7 @@ class SnapshotSession:
         data: "SerializedData",
         index: "SnapshotIndex",
     ) -> None:
-        snapshot_location = extension.get_location(
-            test_location=test_location, index=index
-        )
-        key = (extension.__class__, snapshot_location)
+        key = self._snapshot_write_queue_key(extension, test_location, index)
         queue = self._queued_snapshot_writes.get(key, [])
         queue.append((data, test_location, index))
         self._queued_snapshot_writes[key] = queue
@@ -92,6 +100,27 @@ class SnapshotSession:
                     snapshot_location=snapshot_location, snapshots=queued_write
                 )
         self._queued_snapshot_writes = {}
+
+    def recall_snapshot(
+        self,
+        extension: "AbstractSyrupyExtension",
+        test_location: "PyTestLocation",
+        index: "SnapshotIndex",
+    ) -> Optional["SerializedData"]:
+        """Find the current value of the snapshot, for this session, either a pending write or the actual snapshot."""
+
+        key = self._snapshot_write_queue_key(extension, test_location, index)
+        queue = self._queued_snapshot_writes.get(key)
+        if queue:
+            # find the last (i.e. most recent) write to this index/location in the queue:
+            for queue_data, queue_test_location, queue_index in reversed(queue):
+                if queue_index == index and queue_test_location == test_location:
+                    return queue_data
+
+        # no queue, or no matching write, so just read the snapshot directly:
+        return extension.read_snapshot(
+            test_location=test_location, index=index, session_id=str(id(self))
+        )
 
     @property
     def update_snapshots(self) -> bool:

--- a/src/syrupy/session.py
+++ b/src/syrupy/session.py
@@ -46,6 +46,10 @@ class ItemStatus(Enum):
     SKIPPED = "skipped"
 
 
+_QueuedWriteExtensionKey = Tuple[Type["AbstractSyrupyExtension"], str]
+_QueuedWriteTestLocationKey = Tuple["PyTestLocation", "SnapshotIndex"]
+
+
 @dataclass
 class SnapshotSession:
     pytest_session: "pytest.Session"
@@ -63,8 +67,8 @@ class SnapshotSession:
     )
 
     _queued_snapshot_writes: Dict[
-        Tuple[Type["AbstractSyrupyExtension"], str],
-        Dict[Tuple["PyTestLocation", "SnapshotIndex"], "SerializedData"],
+        _QueuedWriteExtensionKey,
+        Dict[_QueuedWriteTestLocationKey, "SerializedData"],
     ] = field(default_factory=dict)
 
     def _snapshot_write_queue_key(
@@ -72,7 +76,7 @@ class SnapshotSession:
         extension: "AbstractSyrupyExtension",
         test_location: "PyTestLocation",
         index: "SnapshotIndex",
-    ) -> Tuple[Type["AbstractSyrupyExtension"], str]:
+    ) -> _QueuedWriteExtensionKey:
         snapshot_location = extension.get_location(
             test_location=test_location, index=index
         )

--- a/src/syrupy/session.py
+++ b/src/syrupy/session.py
@@ -71,16 +71,16 @@ class SnapshotSession:
         Dict[_QueuedWriteTestLocationKey, "SerializedData"],
     ] = field(default_factory=dict)
 
-    def _snapshot_write_queue_key(
+    def _snapshot_write_queue_keys(
         self,
         extension: "AbstractSyrupyExtension",
         test_location: "PyTestLocation",
         index: "SnapshotIndex",
-    ) -> _QueuedWriteExtensionKey:
+    ) -> Tuple[_QueuedWriteExtensionKey, _QueuedWriteTestLocationKey]:
         snapshot_location = extension.get_location(
             test_location=test_location, index=index
         )
-        return (extension.__class__, snapshot_location)
+        return (extension.__class__, snapshot_location), (test_location, index)
 
     def queue_snapshot_write(
         self,
@@ -89,10 +89,12 @@ class SnapshotSession:
         data: "SerializedData",
         index: "SnapshotIndex",
     ) -> None:
-        key = self._snapshot_write_queue_key(extension, test_location, index)
-        queue = self._queued_snapshot_writes.get(key, {})
-        queue[(test_location, index)] = data
-        self._queued_snapshot_writes[key] = queue
+        ext_key, loc_key = self._snapshot_write_queue_keys(
+            extension, test_location, index
+        )
+        queue = self._queued_snapshot_writes.get(ext_key, {})
+        queue[loc_key] = data
+        self._queued_snapshot_writes[ext_key] = queue
 
     def flush_snapshot_write_queue(self) -> None:
         for (
@@ -117,9 +119,11 @@ class SnapshotSession:
     ) -> Optional["SerializedData"]:
         """Find the current value of the snapshot, for this session, either a pending write or the actual snapshot."""
 
-        key = self._snapshot_write_queue_key(extension, test_location, index)
-        queue = self._queued_snapshot_writes.get(key, {})
-        data = queue.get((test_location, index))
+        ext_key, loc_key = self._snapshot_write_queue_keys(
+            extension, test_location, index
+        )
+        queue = self._queued_snapshot_writes.get(ext_key, {})
+        data = queue.get(loc_key)
         if data is not None:
             return data
 

--- a/tests/integration/test_snapshot_diff.py
+++ b/tests/integration/test_snapshot_diff.py
@@ -1,0 +1,56 @@
+import pytest
+
+_TEST = """
+def test_foo(snapshot):
+    assert {**base} == snapshot(name="a")
+    assert {**base, **extra} == snapshot(name="b", diff="a")
+"""
+
+
+def _make_file(testdir, base, extra):
+    testdir.makepyfile(
+        test_file="\n\n".join([f"base = {base!r}", f"extra = {extra!r}", _TEST])
+    )
+
+
+def _run_test(testdir, base, extra, expected_update_lines):
+    _make_file(testdir, base=base, extra=extra)
+
+    # Run with --snapshot-update, to generate/update snapshots:
+    result = testdir.runpytest(
+        "-v",
+        "--snapshot-update",
+    )
+    result.stdout.re_match_lines((expected_update_lines,))
+    assert result.ret == 0
+
+    # Run without --snapshot-update, to validate the snapshots are actually up-to-date
+    result = testdir.runpytest("-v")
+    result.stdout.re_match_lines((r"2 snapshots passed\.",))
+    assert result.ret == 0
+
+
+def test_diff_lifecycle(testdir) -> pytest.Testdir:
+    # first: create both snapshots completely from scratch
+    _run_test(
+        testdir,
+        base={"A": 1},
+        extra={"X": 10},
+        expected_update_lines=r"2 snapshots generated\.",
+    )
+
+    # second: edit the base data, to change the data for both snapshots (only changes the serialized output for the base snapshot `a`).
+    _run_test(
+        testdir,
+        base={"A": 1, "B": 2},
+        extra={"X": 10},
+        expected_update_lines=r"1 snapshot passed. 1 snapshot updated\.",
+    )
+
+    # third: edit just the extra data (only changes the serialized output for the diff snapshot `b`)
+    _run_test(
+        testdir,
+        base={"A": 1, "B": 2},
+        extra={"X": 10, "Y": 20},
+        expected_update_lines=r"1 snapshot passed. 1 snapshot updated\.",
+    )

